homeassistant/components/heos/media_player.py

```diff
@@ -235,10 +302,26 @@ async def async_update(self):
         controls = self._player.now_playing_media.supported_controls
         current_support = [CONTROL_TO_SUPPORT[control] for control in controls]
         self._supported_features = reduce(ior, current_support, BASE_SUPPORTED_FEATURES)
+        self._group = await self.async_get_group()
```

Comment:
The logic to retrieve group definitions is controller-scoped (not player-scoped) and should follow the same [pattern used to maintain sources](https://github.com/home-assistant/core/blob/dev/homeassistant/components/heos/__init__.py#L220-L344) by introducing a `group_manager` stored in `self.hass.data[HEOS_DOMAIN][DATA_GROUP_MANAGER]`. This manager should subscribe to `EVENT_GROUPS_CHANGED` and trigger the state updates in the dependent entities. It should also encapsulate all of the logic to resolve HEOS `player_ids` to HASS `entity_ids`.

Reply:
Good point. As you have suggested, I have moved the code to the `GroupManager` class which handles the fetching of group information for all players based on events.